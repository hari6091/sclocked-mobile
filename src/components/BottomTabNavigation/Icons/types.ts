import { SvgProps } from 'react-native-svg';

export interface IconProps extends SvgProps {
  focused: boolean;
  color?: string;
  size?: number;
}
