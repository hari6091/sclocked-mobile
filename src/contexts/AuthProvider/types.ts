interface ILogin {
  email: string;
  password: string;
}

export interface IUser {
  email?: string;
  username?: string;
  token?: string;
}

export interface IContext extends IUser {
  signIn: ({ email, password }: ILogin) => Promise<void>;
  logout: () => void;
}

export interface IAuthProvider {
  children: JSX.Element;
}
