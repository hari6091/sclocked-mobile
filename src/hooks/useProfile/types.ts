export interface MyUser {
  uuid?: string;
  id?: string;
  name: string;
  tags: string;
  matricula: string;
  email: string;
  disciplinaOUcargo: string;
  role: string;
  createdAt: string;
}

export interface MyUserSalas {
  uuid?: string;
  id?: string;
  name: string;
  numero: string;
  grupo: string;
  status: string;
  users: MyUser[];
}

export type IUserUpdate = {
  id?: string | undefined;
  name: string | undefined;
  tags: string | undefined;
  matricula: string | undefined;
  disciplinaOUcargo: string | undefined;
  email: string | undefined;
  password: string;
  confPassword: string;
  role: string | undefined;
};
