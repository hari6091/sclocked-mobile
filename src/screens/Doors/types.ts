import { screens } from "../../constants";
import { MainTabScreenProps } from "../../routes/types";

export type DoorsScreenProps = MainTabScreenProps<screens.DOOR>;

export interface ISalaOpt {
  salaNumber: string;
  status: string;
}
