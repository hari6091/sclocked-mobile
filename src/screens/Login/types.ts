import { screens } from "../../constants";
import { RootStackScreenProps } from "../../routes/types";

export type LoginScreenProps = RootStackScreenProps<screens.LOGIN>;

export type SignInType = {
  email: string;
  password: string;
};
