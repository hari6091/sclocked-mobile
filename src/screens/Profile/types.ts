import { screens } from "../../constants";
import { MainTabScreenProps } from "../../routes/types";

export type ProfileScreenProps = MainTabScreenProps<screens.PROFILE>;
