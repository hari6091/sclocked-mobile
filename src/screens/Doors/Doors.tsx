import {
  AlertDialog,
  Box,
  Button,
  Center,
  FlatList,
  Flex,
  HStack,
  Icon,
  Pressable,
  Text,
  VStack,
  useDisclose,
} from "native-base";
import React, { useCallback, useEffect, useState } from "react";
import { MaterialCommunityIcons } from "@expo/vector-icons";
import { DoorsScreenProps, ISalaOpt } from "./types";
import { useProfile, useSalaAuth } from "../../hooks";
import { UserHeader } from "../../components";
import { MyUserSalas } from "../../hooks/useProfile/types";
import { useFocusEffect } from "@react-navigation/native";

const Doors = ({ navigation }: DoorsScreenProps) => {
  const { allUserSalas } = useProfile();
  const { openSala } = useSalaAuth();

  const [salas, setSalas] = useState<MyUserSalas[]>();
  const [isLoading, setIsLoading] = useState<boolean>(false);

  const loadSalas = useCallback(async () => {
    try {
      setIsLoading(true);
      const getSalas = await allUserSalas();
      setSalas(getSalas);
    } finally {
      setIsLoading(false);
    }
  }, [navigation]);

  useEffect(() => {
    loadSalas();
  }, [navigation]);

  useFocusEffect(
    useCallback(() => {
      loadSalas();
    }, [loadSalas])
  );

  const [salaOpt, setSalaOpt] = useState<ISalaOpt | undefined>(undefined);
  const { isOpen, onOpen, onClose } = useDisclose();

  const cancelRef = React.useRef(null);

  const handleAlterSalaState = async (salaOpt: ISalaOpt | undefined) => {
    if (salaOpt) {
      await openSala({ message: salaOpt.salaNumber });
    }
  };

  const renderItem = useCallback(
    ({ item }: { item: MyUserSalas }) => (
      <Pressable
        w="100%"
        onPress={() => {
          setSalaOpt({ salaNumber: item.numero, status: item.status });
          onOpen();
        }}
      >
        <HStack
          w="100%"
          bg="#28374D"
          borderRadius="12px"
          p="16px"
          alignItems="center"
          justifyContent="space-between"
        >
          <VStack maxW="80%">
            <Text fontWeight={700} fontSize="20px" color="#FFFFFF">
              {item?.name + " / " + item.numero ?? "Carregando..."}
            </Text>
            <HStack alignItems="center" space="12px">
              <Text fontSize="16px" color="#FFFFFF">
                {item?.status.toUpperCase() ?? "Carregando..."}
              </Text>
              <Flex
                w="15px"
                h="15px"
                borderRadius="15px"
                bg={
                  item?.status.toUpperCase() === "ATIVO" ? "#00FF29" : "#FF0000"
                }
              />
            </HStack>
          </VStack>
          <Icon
            as={MaterialCommunityIcons}
            name="door"
            size="12"
            color="#FFFFFF"
          />
        </HStack>
      </Pressable>
    ),
    []
  );

  const keyExtractor = useCallback(
    (item: MyUserSalas, index: number) => index.toString(),
    []
  );

  const ItemSeparatorComponent = useCallback(() => <Box mt="22px" />, []);

  return (
    <>
      <UserHeader />
      <Center flex={1} flexDirection="column" safeArea px="10" bg="#FFFFFF">
        <Text fontSize="24px" color="#28374D" mb="12px">
          Salas que vocÃª tem acesso
        </Text>

        <FlatList
          testID="feed-items"
          flex={1}
          data={salas}
          refreshing={isLoading}
          onRefresh={() => loadSalas()}
          renderItem={renderItem}
          keyExtractor={keyExtractor}
          ItemSeparatorComponent={ItemSeparatorComponent}
          onEndReachedThreshold={2}
          initialNumToRender={10}
          w="full"
        />
      </Center>

      <AlertDialog
        leastDestructiveRef={cancelRef}
        isOpen={isOpen}
        onClose={onClose}
      >
        <AlertDialog.Content>
          <AlertDialog.CloseButton />
          <AlertDialog.Header>
            <Text bold fontSize="16px">
              {salaOpt?.status === "inativo" ? "Abrir" : "Fechar"}
              {" sala " + salaOpt?.salaNumber + "?"}
            </Text>
          </AlertDialog.Header>
          <AlertDialog.Body>
            <Text>
              Isto vai alterar o estado da sala para{" "}
              {salaOpt?.status === "inativo" ? "aberta" : "fechada"}
            </Text>
          </AlertDialog.Body>
          <AlertDialog.Footer>
            <Button.Group space={2}>
              <Button
                variant="unstyled"
                colorScheme="coolGray"
                onPress={onClose}
                ref={cancelRef}
              >
                Cancelar
              </Button>
              <Button
                colorScheme={
                  salaOpt?.status === "inativo" ? "emerald" : "coolGray"
                }
                onPress={() => {
                  handleAlterSalaState(salaOpt);
                  onClose();
                }}
              >
                {salaOpt?.status === "inativo" ? "Abrir Sala" : "Fechar sala"}
              </Button>
            </Button.Group>
          </AlertDialog.Footer>
        </AlertDialog.Content>
      </AlertDialog>
    </>
  );
};

export default Doors;
