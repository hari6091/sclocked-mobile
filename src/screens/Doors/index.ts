export { default as Doors } from "./Doors";
