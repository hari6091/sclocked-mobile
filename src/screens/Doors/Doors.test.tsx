import React from "react";
import Doors from "./Doors";
import { fireEvent, render, waitFor } from "../../test/test-utils";
import { DoorsScreenProps } from "./types";

jest.mock("../../hooks/useProfile", () => ({
  useProfile: () => ({
    allUserSalas: jest.fn().mockResolvedValue([
      {
        numero: "101",
        name: "Sala 101",
        status: "ativo",
      },
      {
        numero: "102",
        name: "Sala 102",
        status: "inativo",
      },
    ]),
  }),
}));

jest.mock("../../hooks/useSalaAuth", () => ({
  useSalaAuth: () => ({
    openSala: jest.fn().mockResolvedValue(undefined),
  }),
}));

const navigation = {
  setOptions: jest.fn(),
} as unknown as DoorsScreenProps["navigation"];

const route = {
  params: undefined,
} as unknown as DoorsScreenProps["route"];

describe("Doors screen", () => {
  it("renders the list of salas correctly", async () => {
    const { getByTestId, getByText, toJSON } = render(
      <Doors navigation={navigation} route={route} />
    );

    await waitFor(() => getByTestId("sala-items"));

    expect(getByText("Sala 101 / 101")).toBeTruthy();
    expect(getByText("ATIVO")).toBeTruthy();

    expect(getByText("Sala 102 / 102")).toBeTruthy();
    expect(getByText("INATIVO")).toBeTruthy();

    expect(toJSON()).toMatchSnapshot();
  });

  it("opens the alert when a sala is pressed and handles opening sala", async () => {
    const { getByTestId, getByText, queryByTestId } = render(
      <Doors navigation={navigation} route={route} />
    );
    await waitFor(() => getByTestId("sala-items"));

    fireEvent.press(getByText("Sala 101 / 101"));

    expect(getByText("Fechar sala 101?")).toBeTruthy();
    expect(
      getByText("Isto vai alterar o estado da sala para fechada")
    ).toBeTruthy();

    fireEvent.press(getByTestId("button-alert"));

    await waitFor(() => expect(queryByTestId("alert")).toBeNull());
  });

  test("opens the alert when a sala is pressed and handles closing sala", async () => {
    const { getByTestId, getByText, queryByTestId } = render(
      <Doors navigation={navigation} route={route} />
    );

    await waitFor(() => getByTestId("sala-items"));

    fireEvent.press(getByText("Sala 102 / 102"));

    expect(getByText("Abrir sala 102?")).toBeTruthy();
    expect(
      getByText("Isto vai alterar o estado da sala para aberta")
    ).toBeTruthy();

    fireEvent.press(getByTestId("button-alert"));

    await waitFor(() => expect(queryByTestId("alert")).toBeNull());
  });
});
