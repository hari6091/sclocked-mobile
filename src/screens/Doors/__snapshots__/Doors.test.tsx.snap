// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Doors screen renders the list of salas correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={{}}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "flexDirection": "row",
        "height": "17%",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <Text
      allowFontScaling={false}
      dataSet={{}}
      selectable={false}
      style={
        [
          {
            "color": undefined,
            "fontSize": 12,
          },
          {
            "color": "#28374D",
            "fontSize": 48,
            "height": 48,
            "lineHeight": 48,
            "width": 48,
          },
          {
            "fontFamily": "FontAwesome",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      
    </Text>
    <View
      dataSet={{}}
      style={
        {
          "width": 22,
        }
      }
    />
    <View
      dataSet={{}}
      style={{}}
    >
      <Text
        dataSet={{}}
        style={
          {
            "backgroundColor": undefined,
            "color": "#28374D",
            "fontFamily": undefined,
            "fontSize": 18,
            "fontStyle": "normal",
            "fontWeight": "400",
            "letterSpacing": 0,
            "lineHeight": 27,
            "textAlign": "center",
            "textDecorationLine": undefined,
          }
        }
        testID="username"
      >
        Carregando...
      </Text>
    </View>
  </View>
  <View
    dataSet={{}}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "display": "flex",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "center",
        "paddingLeft": 40,
        "paddingRight": 40,
      }
    }
  >
    <Text
      dataSet={{}}
      style={
        {
          "backgroundColor": undefined,
          "color": "#28374D",
          "fontFamily": undefined,
          "fontSize": 24,
          "fontStyle": "normal",
          "fontWeight": "400",
          "letterSpacing": 0,
          "lineHeight": 36,
          "marginBottom": 12,
          "textDecorationLine": undefined,
        }
      }
      testID="screen-title"
    >
      Salas que você tem acesso
    </Text>
    <RCTScrollView
      ItemSeparatorComponent={[Function]}
      contentContainerStyle={
        [
          {},
          {
            "dataSet": {},
          },
          {},
        ]
      }
      data={
        [
          {
            "name": "Sala 101",
            "numero": "101",
            "status": "ativo",
          },
          {
            "name": "Sala 102",
            "numero": "102",
            "status": "inativo",
          },
        ]
      }
      dataSet={{}}
      getItem={[Function]}
      getItemCount={[Function]}
      initialNumToRender={10}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onRefresh={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <RefreshControlMock
          onRefresh={[Function]}
          refreshing={false}
        />
      }
      refreshing={false}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={[]}
      style={
        {
          "flex": 1,
          "width": "100%",
        }
      }
      testID="sala-items"
      viewabilityConfigCallbackPairs={[]}
    >
      <RCTRefreshControl />
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            dataSet={{}}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "width": "100%",
              }
            }
          >
            <View
              dataSet={{}}
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#28374D",
                  "borderRadius": 12,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "paddingBottom": 16,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "paddingTop": 16,
                  "width": "100%",
                }
              }
            >
              <View
                dataSet={{}}
                style={
                  {
                    "maxWidth": "80%",
                  }
                }
              >
                <Text
                  dataSet={{}}
                  style={
                    {
                      "backgroundColor": undefined,
                      "color": "#FFFFFF",
                      "fontFamily": undefined,
                      "fontSize": 20,
                      "fontStyle": "normal",
                      "fontWeight": "700",
                      "letterSpacing": 0,
                      "lineHeight": 30,
                      "textDecorationLine": undefined,
                    }
                  }
                  testID="sala-title"
                >
                  Sala 101 / 101
                </Text>
                <View
                  dataSet={{}}
                  style={{}}
                />
                <View
                  dataSet={{}}
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                    }
                  }
                >
                  <Text
                    dataSet={{}}
                    style={
                      {
                        "backgroundColor": undefined,
                        "color": "#FFFFFF",
                        "fontFamily": undefined,
                        "fontSize": 16,
                        "fontStyle": "normal",
                        "fontWeight": "400",
                        "letterSpacing": 0,
                        "lineHeight": 24,
                        "textDecorationLine": undefined,
                      }
                    }
                    testID="sala-status"
                  >
                    ATIVO
                  </Text>
                  <View
                    dataSet={{}}
                    style={
                      {
                        "width": 12,
                      }
                    }
                  />
                  <View
                    dataSet={{}}
                    style={
                      {
                        "backgroundColor": "#00FF29",
                        "borderRadius": 15,
                        "display": "flex",
                        "flexDirection": "column",
                        "height": 15,
                        "width": 15,
                      }
                    }
                  />
                </View>
              </View>
              <View
                dataSet={{}}
                style={{}}
              />
              <Text
                allowFontScaling={false}
                dataSet={{}}
                selectable={false}
                style={
                  [
                    {
                      "color": undefined,
                      "fontSize": 12,
                    },
                    {
                      "color": "#FFFFFF",
                      "fontSize": 48,
                      "height": 48,
                      "lineHeight": 48,
                      "width": 48,
                    },
                    {
                      "fontFamily": "material-community",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                󰠚
              </Text>
            </View>
          </View>
          <View
            dataSet={{}}
            style={
              {
                "marginTop": 22,
              }
            }
          />
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            dataSet={{}}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "width": "100%",
              }
            }
          >
            <View
              dataSet={{}}
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#28374D",
                  "borderRadius": 12,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "paddingBottom": 16,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "paddingTop": 16,
                  "width": "100%",
                }
              }
            >
              <View
                dataSet={{}}
                style={
                  {
                    "maxWidth": "80%",
                  }
                }
              >
                <Text
                  dataSet={{}}
                  style={
                    {
                      "backgroundColor": undefined,
                      "color": "#FFFFFF",
                      "fontFamily": undefined,
                      "fontSize": 20,
                      "fontStyle": "normal",
                      "fontWeight": "700",
                      "letterSpacing": 0,
                      "lineHeight": 30,
                      "textDecorationLine": undefined,
                    }
                  }
                  testID="sala-title"
                >
                  Sala 102 / 102
                </Text>
                <View
                  dataSet={{}}
                  style={{}}
                />
                <View
                  dataSet={{}}
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                    }
                  }
                >
                  <Text
                    dataSet={{}}
                    style={
                      {
                        "backgroundColor": undefined,
                        "color": "#FFFFFF",
                        "fontFamily": undefined,
                        "fontSize": 16,
                        "fontStyle": "normal",
                        "fontWeight": "400",
                        "letterSpacing": 0,
                        "lineHeight": 24,
                        "textDecorationLine": undefined,
                      }
                    }
                    testID="sala-status"
                  >
                    INATIVO
                  </Text>
                  <View
                    dataSet={{}}
                    style={
                      {
                        "width": 12,
                      }
                    }
                  />
                  <View
                    dataSet={{}}
                    style={
                      {
                        "backgroundColor": "#FF0000",
                        "borderRadius": 15,
                        "display": "flex",
                        "flexDirection": "column",
                        "height": 15,
                        "width": 15,
                      }
                    }
                  />
                </View>
              </View>
              <View
                dataSet={{}}
                style={{}}
              />
              <Text
                allowFontScaling={false}
                dataSet={{}}
                selectable={false}
                style={
                  [
                    {
                      "color": undefined,
                      "fontSize": 12,
                    },
                    {
                      "color": "#FFFFFF",
                      "fontSize": 48,
                      "height": 48,
                      "lineHeight": 48,
                      "width": 48,
                    },
                    {
                      "fontFamily": "material-community",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                󰠚
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RNCSafeAreaProvider>
`;
